No SystemContextMenu when readOnly is true

justinmc opened the pull request:
When readOnly is true, there is no TextInputConnection, so the SystemContextMenu can't be shown. Instead, this shows the Flutter-drawn context menu.

This video shows the change. You can slightly tell that the system context menu is shown when readOnly is false and the Flutter-drawn context menu is shown when it is true.


https://github.com/user-attachments/assets/91480fa4-cce6-4d63-ae11-df72a229da73


This is the same root cause as https://github.com/flutter/flutter/pull/169238.

Fixes https://github.com/flutter/flutter/issues/170521

justinmc committed:
Show Flutter menu when read only

justinmc committed:
Every other relevant default context menu builder

justinmc committed:
Test for TextField

justinmc committed:
Tests for everything else, and readOnly for SearchBar

justinmc committed:
Skip kIsWeb

justinmc committed:
Missed one [intended]

Renzo-Olivares commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -815,7 +815,9 @@ class CupertinoTextField extends StatefulWidget {
     EditableTextState editableTextState,
   ) {
-    if (defaultTargetPlatform == TargetPlatform.iOS && SystemContextMenu.isSupported(context)) {
+    if (defaultTargetPlatform == TargetPlatform.iOS &&
```
I wonder if we should share this logic somewhere, maybe under `EditableTextState` as `supportsSystemContextMenu`? Not required for this PR but just a thought I had since I see the same logic duplicated across multiple files.

Renzo-Olivares commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -815,7 +815,9 @@ class CupertinoTextField extends StatefulWidget {
-    if (defaultTargetPlatform == TargetPlatform.iOS && SystemContextMenu.isSupported(context)) {
+    if (defaultTargetPlatform == TargetPlatform.iOS &&
+        SystemContextMenu.isSupported(context) &&
+        !editableTextState.widget.readOnly) {
```
It might be helpful to have a comment here on why the system context menu is not supported with `readOnly` set to true.

justinmc replied:
Added on SystemContextMenu.isSupportedByField.

alex-medinsh commented:
>there is no TextInputConnection, so the SystemContextMenu can't be shown

Could you please provide some details on what is blocking us from implementing native context menu without a text connection? I am not very familiar with native iOS development. 

It seems to me that having a native context menu in all cases would be even more important with the release of iOS 26, since IIUC it changes the appearance of the context menu, while Flutter's context menu still resembles the iOS 18 one.

justinmc committed:
isSupportedByField abstraction

hellohuanlin commented:
> @hellohuanlin What would it take to be able to show the system context menu without an active UITextView? Originally we used it to determine which buttons to show, but now that can be specified by the user, so maybe it's easier to do now.

The native edit menu is powered by an `UIEditMenuInteraction`, which needs to be added to a UIView (for our case, it's the backing `UITextInputView`). 

So to support the non-editable text (which doesn't have a backing `UITextInputView`, we would have to add it to other views on screen (e.g. `FlutterView`). We should think about how to reuse the code. It won't be a trivial change, but should be doable.

justinmc commented:
Thanks @hellohuanlin, I created this issue to track: https://github.com/flutter/flutter/issues/171732

justinmc committed:
Analyzer

justinmc committed:
Explain the lack of readOnly support

Renzo-Olivares commented on packages/flutter/lib/src/material/text_form_field.dart:
```diff
@@ -332,7 +332,9 @@ class TextFormField extends FormField<String> {
     EditableTextState editableTextState,
   ) {
-    if (defaultTargetPlatform == TargetPlatform.iOS && SystemContextMenu.isSupported(context)) {
+    if (defaultTargetPlatform == TargetPlatform.iOS &&
```
this should be changed to `SystemContextMenu.isSupportedByField(editableTextState)` I think.

justinmc committed:
Intended reasons

justinmc committed:
Missed one

justinmc committed:
Analyzer fix

I think I had misread the previous analyzer failure as coming from the
wrong file.

Renzo-Olivares commented on packages/flutter/lib/src/material/text_form_field.dart:
```diff
@@ -332,7 +332,7 @@ class TextFormField extends FormField<String> {
     BuildContext context,
     EditableTextState editableTextState,
   ) {
-    if (defaultTargetPlatform == TargetPlatform.iOS && SystemContextMenu.isSupported(context)) {
```
Looks like analyzer is complaining we no longer need the foundation import `package:flutter/foundation.dart` in this file.

Renzo-Olivares reviewed (approved):
LGTM, thanks for writing all those tests! The analyzer is still failing, I think the semantic errors could be cleared up with a rebase.

justinmc committed:
Analyzer

I really messed up fixing this analyzer error, not sure what I misread.

justinmc committed:
Merge branch 'master' into read-only-context-menu

justinmc committed:
Formatting

justinmc committed:
Fix leak

justinmc committed:
Other leaks.

justinmc committed:
Controller used after disposed

justinmc committed:
Forgot to uncomment...

justinmc committed:
Merge branch 'master' into read-only-context-menu

justinmc committed:
Format what the analyzer says to format.

My autoformatter does something different! Not sure what's up.

justinmc committed:
More contra-autoformatter formatting

justinmc committed:
Merge branch 'master' into read-only-context-menu

hasanmhallak commented:
when would this land on stable?
